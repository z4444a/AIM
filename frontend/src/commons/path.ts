export enum Path {
  DEFAULT = '/',
  LOGIN = '/login',
  REQUEST_CREATE = '/requests/create',
  REQUESTS = '/requests',
  REQUEST_CARD = '/requests/:id',
  RESOURCES = '/resources',
  RESOURCE_TYPE_CREATE = '/resource-types/create',
  RESOURCE_TYPE_UPDATE = '/resource-types/:id',
  RESOURCE_TYPES = '/resource-types',
  RESOURCE_POOL_CHART = '/resource-pools/chart',
  RESOURCE_POOL_CREATE = '/resource-pools/create',
  RESOURCE_POOL_UPDATE = '/resource-pools/:id',
  RESOURCE_POOLS = '/resource-pools',
  EMPLOYEE_CARD = '/employees/:id',
  EMPLOYEES = '/employees',
  SETTINGS = '/settings',
}
