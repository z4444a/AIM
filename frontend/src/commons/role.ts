export enum Role {
  USER = 'USER',
  ADMIN = 'ADMIN',
  POOL_CREATOR = 'POOL_CREATOR',
  POOL_OWNER = 'POOL_OWNER',
}
