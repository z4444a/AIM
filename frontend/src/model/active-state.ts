export const ActiveState = [
  {
    key: '1',
    value: 'Активен',
  },
  {
    key: '2',
    value: 'Не активен',
  },
];
