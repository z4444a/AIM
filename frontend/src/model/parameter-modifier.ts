export enum ParameterModifier {
  POOL_PARAMETER = 1,
  REQUEST_PARAMETER = 2,
  ALLOCATION_PARAMETER = 3,
}
