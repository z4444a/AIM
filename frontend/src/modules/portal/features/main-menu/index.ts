import { push } from 'connected-react-router';
import { connect } from 'react-redux';
import { MainMenu } from './main-menu';
import { withTranslation } from 'react-i18next';
import { RootState } from 'typesafe-actions';

export const mapStateToProps = (state: RootState) => ({
  user: state.tokens ? state.tokens.user : null,
});

export const mapDispatchToProps = (dispatch: Function) => ({
  redirect: (path: string) => dispatch(push(path)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation('common')(MainMenu));
