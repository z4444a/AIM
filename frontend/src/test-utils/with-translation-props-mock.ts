import { WithTranslation } from 'react-i18next';

export function getMockWithTranslationProps(): WithTranslation {
  // eslint-disable-next-line
  const t = ((key: string) => `${key}_i18n`) as any;
  // eslint-disable-next-line
  const i18n = {} as any;
  const props: WithTranslation = {
    tReady: true,
    t,
    i18n,
  };

  return props;
}
