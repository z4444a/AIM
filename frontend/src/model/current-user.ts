import { Role } from '../commons/role';

export interface CurrentUser {
  id: number;
  fullName: string;
  username: string;
  role: Role;
}
