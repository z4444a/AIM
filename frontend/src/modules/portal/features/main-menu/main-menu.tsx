import React, { ReactNode } from 'react';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import { mapDispatchToProps, mapStateToProps } from './index';
import MessageTwoTone from '@material-ui/icons/MessageTwoTone';
import People from '@material-ui/icons/People';
import Settings from '@material-ui/icons/Settings';
import TableChart from '@material-ui/icons/TableChart';
import Work from '@material-ui/icons/Work';
import { Path } from '../../../../commons/path';
import { WithTranslation } from 'react-i18next';
import { Role } from '../../../../commons/role';

export type InternalProps = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps> &
  WithTranslation;

export interface State {
  menuItems: MenuItem[];
}

interface MenuItem {
  key: string;
  text: string;
  icon: JSX.Element;
  path: string;
  allowedRoles: Role[];
}

export class MainMenu extends React.PureComponent<InternalProps, State> {
  public constructor(props: InternalProps) {
    super(props);
    this.state = {
      menuItems: this.getMenuItems(),
    };
  }

  public render(): ReactNode {
    const { user } = this.props;
    const role = user ? user.role : Role.USER;
    const { menuItems } = this.state;
    return (
      <List>
        {menuItems
          .filter(item => item.allowedRoles.indexOf(role) >= 0 || role === Role.ADMIN)
          .map(item => this.renderMenuItem(item))}
      </List>
    );
  }

  private renderMenuItem(menuItem: MenuItem): ReactNode {
    return (
      <ListItem
        id={menuItem.key}
        button
        key={menuItem.key}
        onClick={this.handleNavClick(menuItem.path)}
      >
        <ListItemIcon>{menuItem.icon}</ListItemIcon>
        <ListItemText primary={menuItem.text} />
      </ListItem>
    );
  }

  private handleNavClick = (path: string) => () => {
    this.props.redirect(path);
  };

  private getMenuItems(): MenuItem[] {
    const { t } = this.props;
    return [
      {
        key: 'mainMenuItemRequests',
        text: t('mainMenu.requests'),
        icon: <MessageTwoTone />,
        path: Path.REQUESTS,
        allowedRoles: [Role.USER, Role.POOL_OWNER, Role.POOL_CREATOR, Role.ADMIN],
      },
      {
        key: 'mainMenuItemResourceTypes',
        text: t('mainMenu.resourceTypes'),
        icon: <Work />,
        path: Path.RESOURCE_TYPES,
        allowedRoles: [Role.ADMIN],
      },
      {
        key: 'mainMenuItemResourcePools',
        text: t('mainMenu.resourcePools'),
        icon: <TableChart />,
        path: Path.RESOURCE_POOLS,
        allowedRoles: [Role.POOL_OWNER, Role.POOL_CREATOR],
      },
      {
        key: 'mainMenuItemEmployees',
        text: t('mainMenu.employees'),
        icon: <People />,
        path: Path.EMPLOYEES,
        allowedRoles: [Role.ADMIN],
      },
      {
        key: 'mainMenuItemSettings',
        text: t('mainMenu.settings'),
        icon: <Settings />,
        path: Path.SETTINGS,
        allowedRoles: [Role.ADMIN],
      },
    ];
  }
}
