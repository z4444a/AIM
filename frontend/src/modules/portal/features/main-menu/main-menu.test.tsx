import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import { InternalProps, MainMenu, State } from './main-menu';
import { getMockWithTranslationProps } from '../../../../test-utils/with-translation-props-mock';
import sinon from 'sinon';
import { CurrentUser } from '../../../../model/current-user';
import { Role } from '../../../../commons/role';
import { Path } from '../../../../commons/path';

describe('MainMenu', () => {
  function getComponentProps(userRole: Role): InternalProps {
    const user: CurrentUser = {
      id: 0,
      fullName: 'fullName',
      username: 'name',
      role: userRole,
    };
    const withTranslationProps = getMockWithTranslationProps();
    const props: InternalProps = {
      redirect: sinon.spy(),
      user,
      ...withTranslationProps,
    };

    return props;
  }

  function generateShallowWrapper(userRole: Role) {
    const wrapper = shallow(<MainMenu {...getComponentProps(userRole)} />);
    return wrapper;
  }

  function generateMountWrapper(userRole: Role) {
    const wrapper = mount(<MainMenu {...getComponentProps(userRole)} />);
    return wrapper;
  }

  it('should render correctly', () => {
    const wrapper = generateShallowWrapper(Role.ADMIN);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should redirect to requests page', () => {
    const wrapper = generateMountWrapper(Role.ADMIN);
    const menuItems = (wrapper.state() as State).menuItems;
    const item = menuItems.find(item => item.path === Path.REQUESTS);
    const listItem = wrapper.find(`#${item.key}`).at(0);
    listItem.simulate('click');
    expect(wrapper.props().redirect.calledOnceWith(Path.REQUESTS)).toBe(true);
  });

  it('should redirect to resource types page', () => {
    const wrapper = generateMountWrapper(Role.ADMIN);
    const menuItems = (wrapper.state() as State).menuItems;
    const item = menuItems.find(item => item.path === Path.RESOURCE_TYPES);
    const listItem = wrapper.find(`#${item.key}`).at(0);
    listItem.simulate('click');
    expect(wrapper.props().redirect.calledOnceWith(Path.RESOURCE_TYPES)).toBe(true);
  });

  it('should redirect to resource pools page', () => {
    const wrapper = generateMountWrapper(Role.ADMIN);
    const menuItems = (wrapper.state() as State).menuItems;
    const item = menuItems.find(item => item.path === Path.RESOURCE_POOLS);
    const listItem = wrapper.find(`#${item.key}`).at(0);
    listItem.simulate('click');
    expect(wrapper.props().redirect.calledOnceWith(Path.RESOURCE_POOLS)).toBe(true);
  });

  it('should redirect to employees page', () => {
    const wrapper = generateMountWrapper(Role.ADMIN);
    const menuItems = (wrapper.state() as State).menuItems;
    const item = menuItems.find(item => item.path === Path.EMPLOYEES);
    const listItem = wrapper.find(`#${item.key}`).at(0);
    listItem.simulate('click');
    expect(wrapper.props().redirect.calledOnceWith(Path.EMPLOYEES)).toBe(true);
  });

  it('employees item should not be visible to Role.USER', () => {
    const wrapper = generateMountWrapper(Role.USER);
    const menuItems = (wrapper.state() as State).menuItems;
    const item = menuItems.find(item => item.path === Path.EMPLOYEES);
    const actual = wrapper.exists(`#${item.key}`);
    expect(actual).toBe(false);
  });
});
