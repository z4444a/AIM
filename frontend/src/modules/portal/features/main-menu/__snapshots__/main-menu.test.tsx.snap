// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainMenu should render correctly 1`] = `
<WithStyles(List)>
  <WithStyles(ListItem)
    button={true}
    id="mainMenuItemRequests"
    key="mainMenuItemRequests"
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <pure(MessageTwoToneIcon) />
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      primary="mainMenu.requests_i18n"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    id="mainMenuItemResourceTypes"
    key="mainMenuItemResourceTypes"
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <pure(WorkIcon) />
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      primary="mainMenu.resourceTypes_i18n"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    id="mainMenuItemResourcePools"
    key="mainMenuItemResourcePools"
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <pure(TableChartIcon) />
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      primary="mainMenu.resourcePools_i18n"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    id="mainMenuItemEmployees"
    key="mainMenuItemEmployees"
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <pure(PeopleIcon) />
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      primary="mainMenu.employees_i18n"
    />
  </WithStyles(ListItem)>
  <WithStyles(ListItem)
    button={true}
    id="mainMenuItemSettings"
    key="mainMenuItemSettings"
    onClick={[Function]}
  >
    <WithStyles(ListItemIcon)>
      <pure(SettingsIcon) />
    </WithStyles(ListItemIcon)>
    <WithStyles(ListItemText)
      primary="mainMenu.settings_i18n"
    />
  </WithStyles(ListItem)>
</WithStyles(List)>
`;
